File: packages/elo/src/discovery.test.ts
import { describe, expect, it } from 'vitest';
import type { Activity } from '@indiafoss/model';
import { discoveryDeck, learnAffinity, type RankedActivity } from './index.js';

function talk(
  id: string,
  track: string,
  interest?: 'yes' | 'no',
  disposition: RankedActivity['disposition'] = 'normal',
): RankedActivity {
  const activity: Activity = {
    id,
    title: id,
    type: 'talk',
    trackId: track,
    tags: [track],
    flexible: false,
    speakerIds: [],
    source: 'fixture',
  };
  return { activity, rating: 1200, comparisons: 0, disposition, ...(interest ? { interest } : {}) };
}

describe('local talk discovery', () => {
  it('learns a positive answer without any pairwise comparisons', () => {
    const pool = [
      talk('liked', 'rust', 'yes'),
      talk('unrelated', 'design'),
      talk('similar', 'rust'),
    ];
    const deck = discoveryDeck(pool, learnAffinity(pool, []));
    expect(deck[0]?.activity.id).toBe('similar');
    expect(deck[0]?.because).toContain('track:rust');
    expect(deck.map((r) => r.activity.id)).not.toContain('liked');
  });
  it('gives a must-go stronger affinity, without fabricating comparisons', () => {
    const yes = [talk('a', 'rust', 'yes')];
    const must = [talk('a', 'rust', 'yes', 'must-attend')];
    expect(learnAffinity(must, []).affinity.get('track:rust')).toBeGreaterThan(
      learnAffinity(yes, []).affinity.get('track:rust')!,
    );
    expect(must[0]?.comparisons).toBe(0);
  });
  it('excludes dislikes and cancellations, and undo removes their influence', () => {
    const pool = [
      talk('no', 'rust', 'no', 'not-interested'),
      talk('other', 'design'),
      talk('rust', 'rust'),
    ];
    const before = learnAffinity(pool, []);
    expect(before.affinity.get('track:rust')).toBeLessThan(0);
    expect(discoveryDeck(pool, before).map((r) => r.activity.id)).not.toContain('no');
    pool[0] = talk('no', 'rust');
    expect(learnAffinity(pool, []).affinity.get('track:rust')).toBeUndefined();
    pool[0]!.activity.cancelled = true;
    expect(discoveryDeck(pool, learnAffinity(pool, [])).map((r) => r.activity.id)).not.toContain(
      'no',
    );
  });
  it('samples different tracks during cold start and periodically explores', () => {
    const pool = [
      talk('a', 'rust'),
      talk('b', 'rust'),
      talk('c', 'design'),
      talk('d', 'systems'),
      talk('e', 'rust'),
    ];
    const cold = discoveryDeck(pool, learnAffinity(pool, []));
    expect(new Set(cold.slice(0, 3).map((r) => r.activity.trackId)).size).toBe(3);
    pool.push(talk('liked', 'rust', 'yes'));
    const warm = discoveryDeck(pool, learnAffinity(pool, []));
    expect(warm[2]?.activity.trackId).not.toBe('rust');
    expect(discoveryDeck(pool, learnAffinity(pool, []))).toEqual(warm);
  });
});

it('explores after three saved choices even when only the first card is consumed', () => {
  const pool = [
    talk('liked-1', 'rust', 'yes'),
    talk('liked-2', 'rust', 'yes'),
    talk('liked-3', 'rust', 'yes'),
    talk('a-similar', 'rust'),
    talk('z-different', 'design'),
  ];
  expect(discoveryDeck(pool, learnAffinity(pool, []))[0]?.activity.id).toBe('z-different');
  pool[2] = talk('liked-3', 'rust');
  expect(discoveryDeck(pool, learnAffinity(pool, []))[0]?.activity.trackId).toBe('rust');
});

it('a dislike demotes similar talks even without a positive explanation', () => {
  const pool = [
    talk('disliked', 'rust', 'no', 'not-interested'),
    talk('a-similar', 'rust'),
    talk('z-different', 'design'),
  ];
  expect(discoveryDeck(pool, learnAffinity(pool, []))[0]?.activity.id).toBe('z-different');
});

File: apps/android/native/core/src/test/kotlin/org/indiafoss/companion/core/DiscoveryTest.kt
package org.indiafoss.companion.core

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class DiscoveryTest {
    private val day = "2026-09-26"
    private fun talk(id: String, track: String, start: String = "10:00", end: String = "10:30") =
        Activity(id, id, trackId = track, devroomId = track, locationId = track, tags = listOf(track), start = "${day}T$start:00+05:30", end = "${day}T$end:00+05:30")
    private fun bundle(activities: List<Activity>) = EventBundle("fixture", "Fixture", "Asia/Kolkata", "${day}T08:00:00+05:30", "${day}T18:00:00+05:30", activities = activities)

    @Test
    fun `direct positive choices train discovery without comparisons`() {
        val pool = listOf(RankedActivity(talk("liked", "rust"), interest = "yes"), RankedActivity(talk("unrelated", "design")), RankedActivity(talk("similar", "rust")))
        val model = AffinityModel.learn(pool, emptyList())
        assertEquals(listOf("similar", "unrelated"), model.discoveryDeck(pool).map { it.id })
        assertTrue(model.affinity.getValue("track:rust") > 0)
    }

    @Test
    fun `cold discovery samples tracks and excludes direct dislikes`() {
        val pool = listOf(RankedActivity(talk("a", "rust")), RankedActivity(talk("b", "rust")), RankedActivity(talk("c", "design")), RankedActivity(talk("d", "systems")), RankedActivity(talk("no", "design"), disposition = Disposition.NOT_INTERESTED, interest = "no"))
        val deck = AffinityModel.learn(pool, emptyList()).discoveryDeck(pool)
        assertEquals(2, deck.take(2).map { it.trackId }.distinct().size)
        assertTrue(deck.none { it.id == "no" })
    }

    @Test
    fun `whole track reserves gaps but not another programme in the same room`() {
        val b = bundle(listOf(talk("a", "docs", "10:00", "10:30"), talk("b", "docs", "11:00", "11:30"), talk("gap", "other", "10:30", "11:00"), talk("afternoon", "aosp", "14:00", "14:30").copy(locationId = "docs")))
        val plan = Itinerary.forDay(b, day, { 1200.0 }, { Disposition.NORMAL }, { false }, stayTrackIds = setOf("docs"))
        assertEquals(listOf("a", "b", "afternoon"), plan.map { it.activity.id })
        val conflicts = Itinerary.stayConflicts(b, day, setOf("docs"), dispositionOf = { if (it == "gap") Disposition.MUST_ATTEND else Disposition.NORMAL })
        assertTrue(conflicts.any { it.second.id == "gap" })
    }
    @Test fun `exploration survives recomputing the deck after each answer`() {
        val pool = listOf("liked-1", "liked-2", "liked-3").map { RankedActivity(talk(it, "rust"), interest = "yes") } + listOf(RankedActivity(talk("a-similar", "rust")), RankedActivity(talk("z-different", "design")))
        assertEquals("z-different", AffinityModel.learn(pool, emptyList()).discoveryDeck(pool).first().id)
    }

    @Test fun `negative feedback demotes similar unanswered talks`() {
        val pool = listOf(RankedActivity(talk("no", "rust"), interest = "no", disposition = Disposition.NOT_INTERESTED), RankedActivity(talk("a-similar", "rust")), RankedActivity(talk("z-different", "design")))
        assertEquals("z-different", AffinityModel.learn(pool, emptyList()).discoveryDeck(pool).first().id)
    }

}
